Compile error on Intel C++ Classic Compilers

## Description

Trying to integrate Intel C++ Classic Compilers into [nlohmann json](https://github.com/nlohmann/json)'s CI, I'm getting the following error:

```
/home/runner/work/nlohmann-json/nlohmann-json/test/thirdparty/doctest/doctest.h(1238): internal error: assertion failed at: "exprutil.c", line 5535 in make_cast_rescan_operands

          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(==, " == ", DOCTEST_CMP_EQ) //!OCLINT bitwise operator in conditional
          ^

compilation aborted for /home/runner/work/nlohmann-json/nlohmann-json/test/src/unit-user_defined_input.cpp (code 4)
make[6]: *** [test/CMakeFiles/test-user_defined_input.dir/build.make:76: test/CMakeFiles/test-user_defined_input.dir/src/unit-user_defined_input.cpp.o] Error 4
make[5]: *** [CMakeFiles/Makefile2:1045: test/CMakeFiles/test-user_defined_input.dir/all] Error 2
make[4]: *** [Makefile:146: all] Error 2
make[3]: *** [CMakeFiles/ci_test_intel_icpc.dir/build.make:72: CMakeFiles/ci_test_intel_icpc] Error 2
make[2]: *** [CMakeFiles/Makefile2:580: CMakeFiles/ci_test_intel_icpc.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:587: CMakeFiles/ci_test_intel_icpc.dir/rule] Error 2
make: *** [Makefile:309: ci_test_intel_icpc] Error 2
Error: Process completed with exit code 2.
```

The complete CI logs can be found here: https://github.com/nickaein/nlohmann-json/runs/2479842205?check_suite_focus=true
Note that while the compilation with `icl`/`icpc` compilers (Intel **Classic** C++ compilers for Windows/Linux) fails, it passes for `icx`/`icpx` (Intel **LLVM-based** C++ compilers).

### Steps to reproduce

Here is a minimal example to reproduce the issue:

Compile the following code:
```cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
#include <string>

namespace
{
    TEST_CASE("simple test")
    {
        std::string s = "123";
        CHECK(s[0] == '1');
    }
};
```

with Intel C++ Classic Compiler (`icc`/`icpc`) version  2021.2.0. 

The compiler error can be reproduced on Compiler explorer: https://godbolt.org/z/GKd3E8j6o

Also, to reproduce the issue locally, one can use the following Dockerfile as mentioned here https://github.com/nlohmann/json/pull/2736#issuecomment-829717766 to install the compiler in an Ubuntu image. There is also [oneapi-ci project](https://github.com/oneapi-src/oneapi-ci) provided by Intel which contains instructions for setting up compilers in various environments and OSes.

```Dockerfile
FROM nlohmann/json-ci

RUN wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB && \
    APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1 apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB && \
    rm GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB && \
    add-apt-repository -y "deb https://apt.repos.intel.com/oneapi all main" && \
    apt-get update

RUN apt-get install -y --no-install-recommends \
        intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic
```

Running the image of this Dockerfile, I use the following commands to prepare the environment and build the code. I had to provide some extra flags to get the compiler working. Following instructions in `oneapi` to setup the environment may not need these flags.

```bash
source /opt/intel/oneapi/setvars.sh
icpc -std=c++11 -gnu-prefix=x86_64-linux-gnu- -gxx-name=g++-9 example.cpp
```

This will lead to following warnings and the mentioned error:

```
doctest.h(1238): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(==, " == ", DOCTEST_CMP_EQ) //!OCLINT bitwise operator in conditional
          ^

In file included from a.cpp(2):
doctest.h(1239): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(!=, " != ", DOCTEST_CMP_NE) //!OCLINT bitwise operator in conditional
          ^

In file included from a.cpp(2):
doctest.h(1240): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(>,  " >  ", DOCTEST_CMP_GT) //!OCLINT bitwise operator in conditional
          ^

In file included from a.cpp(2):
doctest.h(1241): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(<,  " <  ", DOCTEST_CMP_LT) //!OCLINT bitwise operator in conditional
          ^

In file included from a.cpp(2):
doctest.h(1242): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(>=, " >= ", DOCTEST_CMP_GE) //!OCLINT bitwise operator in conditional
          ^

In file included from a.cpp(2):
doctest.h(1243): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(<=, " <= ", DOCTEST_CMP_LE) //!OCLINT bitwise operator in conditional
          ^

In file included from a.cpp(2):
doctest.h(1390): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_NOINLINE void binary_assert(const DOCTEST_REF_WRAP(L) lhs,
          ^

In file included from a.cpp(2):
doctest.h(1398): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_NOINLINE void unary_assert(const DOCTEST_REF_WRAP(L) val) {
          ^

doctest.h(1238): internal error: assertion failed at: "exprutil.c", line 5535 in make_cast_rescan_operands

          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(==, " == ", DOCTEST_CMP_EQ) //!OCLINT bitwise operator in conditional
```

### Extra information

* doctest version: **v2.4.6**
* Operating System: Ubuntu 20.04.2
* Compiler+version: **Intel Classic C++ Compiler (icc) 2021.2.0 20210228**

--- a/doctest/doctest.h
+++ b/doctest/doctest.h
@@ -1079,7 +1079,7 @@ DOCTEST_CLANG_SUPPRESS_WARNING_WITH_PUSH("-Wunused-comparison")
 // If not it doesn't find the operator or if the operator at global scope is defined after
 // this template, the template won't be instantiated due to SFINAE. Once the template is not
 // instantiated it can look for global operator using normal conversions.
-#define SFINAE_OP(ret,op) decltype(doctest::detail::declval<L>() op doctest::detail::declval<R>(),static_cast<ret>(0))
+#define SFINAE_OP(ret,op) decltype(doctest::detail::declval<L>() op doctest::detail::declval<R>(),ret{})
 
 #define DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(op, op_str, op_macro)                              \
     template <typename R>                                                                          \
@@ -1108,6 +1108,7 @@ DOCTEST_CLANG_SUPPRESS_WARNING_WITH_PUSH("-Wunused-comparison")
         bool   m_passed;
         String m_decomp;
 
+        Result() = default;
         Result(bool passed, const String& decomposition = String());
 
         // forbidding some expressions based on this table: https://en.cppreference.com/w/cpp/language/operator_precedence

--- a/doctest/parts/doctest_fwd.h
+++ b/doctest/parts/doctest_fwd.h
@@ -1076,7 +1076,7 @@ DOCTEST_CLANG_SUPPRESS_WARNING_WITH_PUSH("-Wunused-comparison")
 // If not it doesn't find the operator or if the operator at global scope is defined after
 // this template, the template won't be instantiated due to SFINAE. Once the template is not
 // instantiated it can look for global operator using normal conversions.
-#define SFINAE_OP(ret,op) decltype(doctest::detail::declval<L>() op doctest::detail::declval<R>(),static_cast<ret>(0))
+#define SFINAE_OP(ret,op) decltype(doctest::detail::declval<L>() op doctest::detail::declval<R>(),ret{})
 
 #define DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(op, op_str, op_macro)                              \
     template <typename R>                                                                          \
@@ -1105,6 +1105,7 @@ DOCTEST_CLANG_SUPPRESS_WARNING_WITH_PUSH("-Wunused-comparison")
         bool   m_passed;
         String m_decomp;
 
+        Result() = default;
         Result(bool passed, const String& decomposition = String());
 
         // forbidding some expressions based on this table: https://en.cppreference.com/w/cpp/language/operator_precedence

